### ISSUE TITLE
`template-indent` rule conflicts with the prettier

### ISSUE THREAD

**Issue by TimShilov** [sindresorhus/eslint-plugin-unicorn#1815]
Hello,
I have encountered a problem with `unicorn/template-indent` rule. It's not indenting the string properly which causes conflict with `eslint-plugin-prettier`. Fixing one results on other complaining.
I know there's an option to specify the amount of spaces to use for indentation but it feels wrong to hard-code number of spaces in multiple configs.
I wonder if there's a way for this rule to either make this rule respect prettier/editorconfig style guide or make it figure out the indent needed based on the other indents inside of the template literal?

Here's an example file:
```js
const AppPanel = {};

const Template = (args, options) => ({
    components: { AppPanel },
    setup() {
        return { args };
    },
    // On the next line I get 'ESLint: Templates should be properly indented. (unicorn/template-indent)'
    // ... even though I think it is indented properly 
    template: /* html */ `
        <div style="width:400px">
            <AppPanel>
                <div style="padding:2.4rem">{{args.default}}</div>
            </AppPanel >
        </div>
    `,
});
```

If I autofix this issue I get this (irrelevant code omitted):
```js
    template: /* html */ `
      <div style="width:400px">
          <AppPanel>
              <div style="padding:2.4rem">{{args.default}}</div>
          </AppPanel>
      </div>
    `,
```
The problem is the first indent is using 2 spaces while the nested nodes are using 4 spaces (like the rest of the file) so Prettier starts complaining because it wants to indent them according to the styleguide (defined in editorconfig).

I think in this case it should use 4 spaces for indentation automatically.

**Comment by fisker** [comment]
It uses 2 spaces by default. You can config it like

```js
{
	'unicorn/template-indent': [
		'error', {
			indent: 4,
		}
	]
}
```

**Comment by TimShilov** [comment]
Yea, I know that. The question is if there's a way to make it work properly without hard-coding the number of spaces?
There's multiple levels of 4 spaces indentation inside the string so I think it is safe to automatically take that as "indent"?
Maybe it is possible to create an option to enable this behavior?

**Comment by fisker** [comment]
What if there is no clue at all?

```js
export default /* html */ `
    <div></div>
`;
```

**Comment by TimShilov** [comment]
That's fair. The more I think about it the more it seems like unnecessary complication. 😅
I'll stick to the option for now then. Thank you!

**Comment by JounQin** [comment]
This rule is in conflict with prettier for `.mdx` files too.

https://github.com/mdx-js/eslint-mdx/blob/master/test/fixtures/380.mdx#L49

@fisker

**Comment by fisker** [comment]
Don't know what's the problem.

**Comment by JounQin** [comment]
```jsx
<Canvas>
  <Story
    name="Basic"
    height="100px"
  >
    {{
      template: /* HTML */ `
        <aui-tab-group [(tab)]="tab">
          <aui-tab
            name="a"
            label="Tab 0"
          >
            <aui-card> Content 1</aui-card>
          </aui-tab>
          <aui-tab
            name="b"
            label="Tab 1"
          >
            <aui-card> Content 2</aui-card>
          </aui-tab>
          <aui-tab
            name="c"
            label="Tab 2"
          >
            <aui-card> Content 3</aui-card>
          </aui-tab>
        </aui-tab-group>
      `,
      props: {
        tab: 'a',
      },
    }}
  </Story>
</Canvas>
```

The `template` throws this lint error.

**Comment by fisker** [comment]
What the rule want it be?

**Comment by fisker** [comment]
I really don't know where is the problem. Can you post diff?

**Comment by JounQin** [comment]
```diff
diff --git a/test/fixtures/380.mdx b/test/fixtures/380.mdx
index 55d7066..712b523 100644
--- a/test/fixtures/380.mdx
+++ b/test/fixtures/380.mdx
@@ -46,6 +46,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-tab-group [(tab)]="tab">
           <aui-tab
             name="a"
@@ -66,11 +67,12 @@ import { LazyTestComponent } from './lazy-test.component'
             <aui-card> Content 3</aui-card>
           </aui-tab>
         </aui-tab-group>
-      `,
+       `,
       props: {
         tab: 'a',
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -83,16 +85,18 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-tab-group [type]="type">
           <aui-tab label="Tab 0"><aui-card> Content 1 </aui-card></aui-tab>
           <aui-tab label="Tab 1"><aui-card> Content 2 </aui-card></aui-tab>
           <aui-tab label="Tab 2"><aui-card> Content 3 </aui-card></aui-tab>
         </aui-tab-group>
-      `,
+       `,
       props: {
         type: TabType.Card,
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -105,6 +109,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-form-item>
           <label auiFormItemLabel>尺寸</label>
           <aui-radio-group
@@ -122,11 +127,12 @@ import { LazyTestComponent } from './lazy-test.component'
           <aui-tab label="Tab 1"><aui-card> Content 2 </aui-card></aui-tab>
           <aui-tab label="Tab 2"><aui-card> Content 3 </aui-card></aui-tab>
         </aui-tab-group>
-      `,
+       `,
       props: {
         size: TabSize.Medium,
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -138,24 +144,26 @@ import { LazyTestComponent } from './lazy-test.component'
     height="200px"
   >
     {{
-      template: /* HTML */ `<aui-tab-group>
-        <ng-container *auiTabTitle>aa2</ng-container>
-        <aui-tab>
-          <ng-container *auiTabLabel>
-            <aui-icon icon="sun"></aui-icon>
-            Custom Label
-          </ng-container>
-          <aui-card> Content 1 </aui-card>
-        </aui-tab>
-        <aui-tab label="Tab 1"><aui-card>Content 2</aui-card></aui-tab>
-        <aui-tab>
-          <ng-container *auiTabLabel>
-            <aui-icon icon="moon"></aui-icon>
-            Custom Label
-          </ng-container>
-          <aui-card> Content 3 </aui-card>
-        </aui-tab>
-      </aui-tab-group>`,
+      template: /* HTML */ `<
+        aui-tab-group>
+                <ng-container *auiTabTitle>aa2</ng-container>
+                <aui-tab>
+                  <ng-container *auiTabLabel>
+                    <aui-icon icon="sun"></aui-icon>
+                    Custom Label
+                  </ng-container>
+                  <aui-card> Content 1 </aui-card>
+                </aui-tab>
+                <aui-tab label="Tab 1"><aui-card>Content 2</aui-card></aui-tab>
+                <aui-tab>
+                  <ng-container *auiTabLabel>
+                    <aui-icon icon="moon"></aui-icon>
+                    Custom Label
+                  </ng-container>
+                  <aui-card> Content 3 </aui-card>
+                </aui-tab>
+              </aui-tab-group
+      >`,
       props: {
         tabs: Array.from({ length: 3 }).fill(),
       },
@@ -174,6 +182,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-tab-group>
           <aui-tab
             *ngFor="let tab of tabs; index as i; count as len"
@@ -193,7 +202,7 @@ import { LazyTestComponent } from './lazy-test.component'
             <aui-icon icon="plus"></aui-icon>
           </button>
         </aui-tab-group>
-      `,
+       `,
       props: {
         tabs: [1, 2, 3],
         add(num = 1) {
@@ -208,6 +217,7 @@ import { LazyTestComponent } from './lazy-test.component'
         },
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -220,6 +230,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-tab-group>
           <aui-tab label="Tab 0"> <aui-card> Content 1 </aui-card> </aui-tab>
           <aui-tab
@@ -234,11 +245,12 @@ import { LazyTestComponent } from './lazy-test.component'
             ><aui-card> Content 4 </aui-card></aui-tab
           >
         </aui-tab-group>
-      `,
+       `,
       props: {
         tabs: Array.from({ length: 10 }).fill(),
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -253,6 +265,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-form-item>
           <label auiFormItemLabel>启用</label>
           <aui-radio-group
@@ -281,12 +294,13 @@ import { LazyTestComponent } from './lazy-test.component'
             >
           </aui-tab>
         </aui-tab-group>
-      `,
+       `,
       props: {
         tabs: Array.from({ length: 10 }).fill(),
         lazy: true,
       },
     }}
+
   </Story>
 </Canvas>
 
@@ -348,6 +362,7 @@ import { LazyTestComponent } from './lazy-test.component'
   >
     {{
       template: /* HTML */ `
+
         <aui-tab-group size="large">
           <ng-container *auiTabTitle>Title</ng-container>
           <aui-tab label="Tab 0"> <aui-card> Content 1 </aui-card> </aui-tab>
@@ -355,8 +370,9 @@ import { LazyTestComponent } from './lazy-test.component'
           <aui-tab label="Tab 2"><aui-card> Content 3 </aui-card></aui-tab>
           <aui-tab label="Tab 3"><aui-card> Content 4 </aui-card></aui-tab>
         </aui-tab-group>
-      `,
+       `,
     }}
+
   </Story>
 </Canvas>
```

**Comment by JounQin** [comment]
If you want, you can debug by toggle https://github.com/mdx-js/eslint-mdx/blob/master/test/fixtures.test.ts#L26 and run `yarn eslint --no-ignore test/fixtures/380.mdx --fix`.

**Comment by JounQin** [comment]
Perhaps it is related to the parsed AST loc itself. I'll check tomorrow.

**Comment by JounQin** [comment]
@fisker OK, I can reproduce in a plain jsx file: https://stackblitz.com/edit/node-w8xpuk?file=index.js

**Comment by JounQin** [comment]
And it is very wired about `node.quasis`'s position:

```log
node.quasis: <ref *1> [
  Node {
    type: 'TemplateElement',
    start: 38,
    end: 102,
    loc: SourceLocation {
      start: [Position],
      end: [Position],
      filename: undefined,
      identifierName: undefined
    },
    range: [ 38, 102 ],
    value: {
      raw: '<div abc>\n      <div>Nested</div>\n      Hello World\n    </div>',
      cooked: '<div abc>\n      <div>Nested</div>\n      Hello World\n    </div>'
    },
    tail: true,
    parent: Node {
      type: 'TemplateLiteral',
      start: 38,
      end: 102,
      loc: [SourceLocation],
      range: [Array],
      expressions: [],
      quasis: [Circular *1],
      parent: [Node]
    }
  }
]
joined: "<div abc>\n      <div>Nested</div>\n      Hello World\n    </div>"
```

But in [astexplorer](https://astexplorer.net/#/gist/aeca7acab05ec0887cf5372aba498b9e/65d27f47ffed55a1234c9cba07f39eeb0175ee8a), it is

```json
{
  "type": "TemplateLiteral",
  "start": 38,
  "end": 102,
  "expressions": [],
  "quasis": [
    {
      "type": "TemplateElement",
      "start": 39,
      "end": 101,
      "value": {
        "raw": "<div abc>\n      <div>Nested</div>\n      Hello World\n    </div>",
        "cooked": "<div abc>\n      <div>Nested</div>\n      Hello World\n    </div>"
      },
      "tail": true
    }
  ]
}
```

Which is same in `remark-mdx` parser using `acorn` under the hood.

**Comment by JounQin** [comment]
@fisker Any plain to fix this https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1815#issuecomment-1126881651? (The next comment about the spec issue is not related)

This reproduction means the empty whitespace leading and tailing should be ignored by this rule.

Its fix is incorrect and unsafe, and ugly:

```jsx
<Story>
  {{
    template: /* HTML */ `
      <div abc>
            <div>Nested</div>
            Hello World
          </div>
    `,
  }}
</Story>;
```

**Comment by fisker** [comment]
I thought about this before, but it's hard to do it right.
